WIP: Gnosis DutchX Buyback

samparsky opened the pull request:
Work in progress

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+        bool allowExternalPoke;
+    }
+
+    // mapping of the sell token to 
```
:) maybe we can complete the comment

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+contract BuyBack {
+
+    address owner;    
+    address dutchXAddress;
```
If we have declared `dx`, do we really need to save also `dutchXAddress`?

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+        require(address(_sellToken) != address(0));
+        require(_sellToken != _buyToken);
+        require(_auctionIndexes.length == _auctionAmounts.length);
+        require(buybacks[_userAddress].sellToken == address(0)); // ensure the user address doesn't exist
```
We only allow a buy back per user? What if a user wants to do a buy back of A-WETH and B-WETH?

samparsky replied:
Yeah they can modify the auction indexes and amount. Since the buyback would execute in an a  uction index.

anxolin replied:
I still don't understand the setup. 

1. I see there's a single owner, does this mean that the contract is used only for one project? Do we need to deploy a contract per owner?

2. I don't understand the parameter `_userAddress`, it looks like it's not an user address, and it's just being used as an `ID`.

3. It's sent a commitment for selling an amount in some auction index, but this commitment won't have any effect if someone wants to execute the postSell order, because the contract may not have enough balance of the token. You just send the amounts, but nothing forces you to add the tokens to satisfy them.

samparsky replied:
1. The idea was that the contract can be used for multiple projects. It's also possible to deploy a contract per owner if needed.

2. It's the address of the user that wants to create the buyback config. It's also used as an id, since each address is unique.

3. Yeah, so when executing a postSellOrder the contract checks for that the user has enough balance to execute the `sellOrder` if not reverts the transaction. Right now nothing forces you to add the tokens to satisfy them but then when executing the `postSellOrder` the contract ensures you have enough balance before posting the `sellOrder` to the gnosis contract.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+    mapping(address => Buyback) buybacks;
+    
+    // mapping of useraddress to auction index
+    mapping(address => mapping(uint => uint)) internal auction;
```
Can we review the comments and or name of the variables. It's not evident to me what are they for. In this case, the name is too general and the type doesn't match the comment

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+            auction[_userAddress][_auctionIndexes[i]] = _auctionAmounts[i];
+        }
+
+        Buyback memory buyback = Buyback(_sellToken, _buyToken, _burnAddress, _burn, _auctionIndexes, _timeInterval, 0, true, _allowExternalPoke);
```
Can we fix this and other lint warnings. This line is too long.

If it has a lot of params, I think is better something like:
```js
Buyback(
  _sellToken,
  _buyToken,
   ...
)
```

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+        buybacks[_userAddress] = buyback;
+        
+        // set default tip price
+        tips[_userAddress] = 1000000000 wei; // 
```
Can we use constants or have it parametrized?

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+        uint newBuyerBalance;
+        uint dxAuctionIndex;
+
+        uint[] auctionIndexes = buybacks[_userAddress].auctionIndexes;
```
`buybacks[_userAddress]` can be factorized. This way we avoid doing several times the same access. This comment applies also to other methods that have this small issue.

I think it also make it more readable.

samparsky replied:
What do you mean by factorized? Factory design pattern?

anxolin replied:
No, I just mean that this block (and other similar ones) can be refactored:

```js
        uint[] auctionIndexes = buybacks[_userAddress].auctionIndexes;
        address sellToken = buybacks[_userAddress].sellToken;
        address buyToken = buybacks[_userAddress].buyToken;
        bool shouldBurnToken = buybacks[_userAddress].shouldBurnToken;
        address burnAddress = buybacks[_userAddress].burnAddress;
```

sth like:
```js
        Buyback buyback =  buybacks[_userAddress]
        uint[] auctionIndexes = buyback.auctionIndexes;
        address sellToken = buyback.sellToken;
        address buyToken = buyback.buyToken;
        bool shouldBurnToken = buyback.shouldBurnToken;
        address burnAddress = buyback.burnAddress;
```

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+
+        bool allowExternalPoke = buybacks[_userAddress].allowExternalPoke;
+        if(allowExternalPoke == false){
+            require(msg.sender == owner);
```
I think we should do all the requires at the beginning. 

For example, we can delay the declaration of this:
```js
        uint newBuyerBalance;
        uint dxAuctionIndex;
```

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,691 @@
+            require(msg.sender == owner);
+        }
+
+        require(isTimePassed(_userAddress) == true); // ensure the min time has passed
```
I don't think we need to compare with `== true` in a require. Can we remove them in all the requires?

anxolin commented on gnosis-buyback/test/buyback.js:
```diff
@@ -0,0 +1,374 @@
+  })
+
+  it("Should allow to modify burn", async() => {
+    await buyBack.modifyBurn(InitAccount, false, {from: InitAccount});
```
I think this and other test should be testing the functionality. In this example you only test that the operation succeeded.

A good pattern to write the test is to follow a GIVEN, WHEN, THEN approach. In this case:
// GIVEN: A buyback with burn enabled
// WHEN: Modify it to disable it
// THEN: The burn is disabled

anxolin replied:
Also, I think is good if we test that only the owner can perform some operations and other especial cases  that is implemented in the contract.

anxolin commented on gnosis-buyback/test/buyback.js:
```diff
@@ -0,0 +1,374 @@
+
+    // deposits ethertoken
+    await buyBack.depositSellToken(InitAccount, 20e18, {from: InitAccount})
+    await buyBack.depositSellToken(InitAccount, 20e18, {from: InitAccount})
```
Why the 2 last lines are equal? it's that ok?

samparsky commented:
@anxolin I have added more details to the readme doc & added more tests

anxolin commented:
Thank you very much @samparsky , this week I'll try to find time to review it. Worst case I think I'll have it review by next wednesday.

samparsky commented:
@anxolin Is there a way to possibly have it reviewed sooner?

samparsky commented:
@anxolin Still waiting on your review :)

anxolin commented:
I'm sorry Sam, we've been very busy this days. We'll review it as soon as we can and get back to you.

Thanks for adding the documentation. We'll review that, and we can organize a call with you after.

samparsky commented:
@anxolin Okay cool, hopefully it doesn't turn over into next year

samparsky commented:
@anxolin Happy new year

anxolin commented on gnosis-buyback/Readme.md:
```diff
@@ -0,0 +1,229 @@
+# DutchX Token Buyback Implementation
```
Thanks, having the README really helps understanding the approach

anxolin commented on gnosis-buyback/package.json:
```diff
@@ -0,0 +1,39 @@
+    "truffle": "^4.1.5"
+  },
+  "scripts": {
+    "test": "sh scripts/test.sh",
```
I executed the tests, and 5 were failing.

samparsky replied:
I will look into this

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+    mapping (address => mapping(address => uint)) internal balances;
+
+    // mapping of user address to ether deposit value
+    mapping(address => uint) internal etherBalance;
```
It looks like you can get funds stuck in the contract? One example is, if you add a tip, but you don't do the post sell order when the time is right.

samparsky replied:
Yeah true, I would add a method to enable withdrawal of ether deposited for tips

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+     * @param _auctionIndexes Auction index the to participate 
+     * @param _auctionAmounts Auction amount to fill in auction index
+     */
+    function modifyAuctionAmountMulti(
```
For this an all the other modification functions: Do we allow to modify a buy back?

The way I understood it was as a commitment for a project. But it looks like a project can add the buy back, and then change his mind.

samparsky replied:
Yeah, I figured a project can want to increase or decrease the buyback amount they want to execute based off parameters external to the contract.

Take for example Binance coin (BNB), Binance executes buyback of BNB based off the profits for the month so enabling the project to modify it accordingly seemed okay.

anxolin replied:
The buy back must be enforced, otherwise loosed it's usability. It should be an unbreakable commitment. 

If you want to modify, it should be to allow to add new buybacks, but users should have the certainty that they'll participate in the auction.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+        uint _auctionIndex, 
+        uint _auctionAmount
+        ) public onlyOwner {
+        require(_auctionAmount > 0);
```
All requires should have a message

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+        // ensure not currently participating in an auction
+        require(alreadyClaimed[_userAddress][acIndex] == false);
+
+        if(buybacks[_userAddress].auctionIndexes.length == 1){
```
Is this an optimization? or why the case of one element is different than the one with N elements?

samparsky replied:
An edge case. Solidity throws when trying to access out of bound index.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+        uint acIndex = dxAuctionIndex[_userAddress][auctionIndex];
+
+        // ensure not currently participating in an auction
+        require(alreadyClaimed[_userAddress][acIndex] == false);
```
I don't understand, shouldn't it be the opposite? require that the auctions was indeed claimed?

samparsky replied:
Yeah, the comment is wrong. Should be `ensure the auctions has been claimed`

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+        address sellToken = userBuyback.sellToken;
+        address buyToken = userBuyback.buyToken;
+
+        require(balances[_userAddress][sellToken] == 0); // ensure all withdrawal is made
```
The code is still full of repetition. Afaik, on top of more readable, is more efficient to factorize. In this case: `balances[_userAddress]` but there's many examples in the code.

This was already mentioned in the previous PR.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+        require(_amount > 0);
+        require(buybacks[_userAddress].sellToken != address(0));
+
+        address sellToken = buybacks[_userAddress].sellToken;
```
Please, try to keep it DRY. In this case this line is almost like the previous one

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+     */
+    function() external payable {
+        address userAddress = msg.sender;
+        etherBalance[userAddress] += msg.value;
```
3 things related to this method:

- Is there a strategy to avoid losing funds in the contract? How do I get them out after we end our commitment?
- How do we know for which tip is which ether? 
- How do we ensure that we have enough ether to pay the tips? You can add a tip, and not the ether, right?

samparsky replied:
The tipping implementation is a soft contract, that is if the user has enough ether for tipping the person poking the `postSellOrder` function is tipped else nothing happens.

Also each project address is mapping to their ether deposit values.

anxolin replied:
- So, not commitment also for the tip?
- Can you answer the other points too, pls

anxolin commented on gnosis-buyback/contracts/CoolAppdependencies.sol:
```diff
@@ -0,0 +1,4 @@
+pragma solidity ^0.4.21;
+import "@gnosis.pm/dx-contracts/contracts/DxDevDependencies.sol";
+
+contract CoolAppDependencies {}
```
Small comment :) 
This CoolApp was just an example in the dx-examples, you should call it with the name of this project.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+            require(balances[_userAddress][sellToken] >=  amount); // check if the user as enough balance
+
+            approveDutchX(sellToken, amount);
+            depositDutchx(sellToken, amount);
```
This two might fail if there's not enough tokens. It should be a commitment.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+    }
+
+    // mapping of the user address to token addresses & balance
+    mapping (address => mapping(address => uint)) internal balances;
```
Why all the attributes are internal? is there any benefit? I mean, this attributes make sense to use them in an interface that uses this contract.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,756 @@
+     * @param _tokenAddress Address of the  token
+     * @param _amount Amount of tokens to burn
+     */
+    function burnTokens(address _tokenAddress, uint _amount) internal {
```
Why do we need this function? Is not the same as `burnTokensWithAddress`?

samparsky commented:
@anxolin Thanks for the review. I have gone through the code and made a number of changes to the logic. The `postSellOrder` function now ensures the user has enough balance to post a `sellOrder`.
Also, I have removed `_userAddress` from all the function signatures as it's not needed.

Also added a function `withdrawEther` to enable withdrawing ether deposit from the contract.

I have added error messages to `require` where needed.

The test cases have been fixed and is working okay now.

Please let me know when you will be available

anxolin commented:
Hi @samparsky ,

I still feel the issues raised in the last review weren't addressed.

You can reply to them point by point if you want.

1. Looks like the contract is meant to be used with a single owner 

2. The parameter `_userAddress`, base of all operations, looks like it's modelling just an ID, for identifying the buy back. Is not an user address 

3. The buy backs are not enforced. The contract allows to modify everything 

4. The buy backs are not enforced. The contract allows to have less tokens than the ones it needs to fulfil the commitment.

 5. The tip is not enforced. Even if you say that you give a tip for poking the contract, you can skip the part of adding ether as a tip.

6. You can get the Ether from tips stuck, if no-one sends a post sell order 

7. Some test don’t pass: 5 of them

8. Some code quality notes, and visibility issue 

samparsky commented:
1. The contract can be used by single or multiple owners

2. The parameter `_userAddress` has been removed.

3. Yeah, the buybacks allow to modify parameters to a buyback. It doesn't mean they aren't enforced. e.g.  A project can decide to change params of a montly buyback due to factors like profit or crypto bear market. 
But then I can remove the modify functions if you aren't convinced.

4. You don't have to have the deposited tokens to add a buyback. A scenario is when a project does monthly buybacks. But when invoking `postSellOrder` it ensures the project has enough token deposit to fulfill the buyback. https://github.com/GnosisEcosystemFund/Gnosis-Bounties-/blob/638ac2447f9778a961fca6774b9c6b305f1ab7f4/gnosis-buyback/contracts/Buyback.sol#L442

5. Yes the tip is not enforced its optional, a project can  decide to tip or not. If they decide to tip it's enforced if not then it's not enforced

6. You can withdraw the Ether for the tips through the `withdrawEther` function
https://github.com/GnosisEcosystemFund/Gnosis-Bounties-/blob/638ac2447f9778a961fca6774b9c6b305f1ab7f4/gnosis-buyback/contracts/Buyback.sol#L593

7. All the test cases pass you might need to update your local branch

8. A number of code changes have been made

samparsky commented:
@anxolin hello, have you been able to check it out?

anxolin commented:
From your answers I get that it doesn't create the commitment.

The commitment is required, because the whole point of this contract is to offer the guarantees to the users that you will buy back in a given auction. You can make a commitment bigger (more stake), but in any case you shouldn't be able, by design, to fail to fulfill it.

samparsky commented:
@anxolin Is it possible to have a 5 mins proper discussion on Gitter on the needed changes and the requirements of the application. This one month back and forth isn't productive and its quite a waste of effort and time.

I understand that you are busy but it would be awesome if we can schedule 5 mins next week Monday to just discuss next steps and appropriate system design

anxolin commented:
Hi Sam, I agree, but I think the requirements are described https://gitcoin.co/issue/GnosisEcosystemFund/Gnosis-Bounties-/3/1626, and I already commented the issues with the current approach you are taking.

The biggest fundamental problem, was commented already in the first review. The contract **should allow to pre-commit the participation in specific auctions**. 
Note that the current contract, doesn't create any commitment that can be enforced, it's just a register of the intention of doing so, and I don't see how this offer any guarantees to the users that want this buy back to happen.

I'm afraid I cannot give you the details on how to do it, nor I have capacity right now to do so. I'm happy to answer concrete questions or a proposal though.

I think you've done already a nice job, I'm positive you can make the changes to make it work

samparsky commented:
@anxolin The changes required have been made. Please kindly review

samparsky commented:
@anxolin Hello its been quite a while

anxolin commented:
I'm sorry @samparsky we are a bit out of capacity, but I'll promise I'll find time for you this week. 

Looking forward to see your solution already!

anxolin commented:
Hei Sam!

I see that you modified the contract in the right direction, but I still have some concerns about the contract related to what I was saying in the previous reviews. 

I'll write my thoughts. Let's see it with some examples:

1. I can create a buy back for an auction that will be next week, people are prepared to participate in that auction because they saw my commitment, *BUT*, right before the auction, I can decide to remove my buy back, so I end up not participating. Basically, I broke my promise.

2. I don't see how the tip is enforced. For example, I say to my users, if you submit my buy back, I'll give you `0.1 Ether`, but I don't have `0.1 Ether` in my account. Furthermore, it would be a easy way for me to break my commitment of participating in the auction. This is because the check is done in `postSellOrder`. So not only I don't pay the Ether to the user submitting the transaction to make me participate, I'll won't participate in the auction I committed to, and on top of that, I'll make this user to loose gas because of the revert of the transaction.

3. I'm not 100% sure what happens when I don't participate in the auction, and no-one enforces me to participate. Can I get the money back of the contract?

4. It feels that I can still change the buy backs as I please. For example I commit to burning the tokens, but then I change my mind and I just call `modifyBurn`, or instead of burning in the address `0x0` I decide to use`modifyBurnAddress` to send the tokens to my account (again, breaking my commitment). I feel this limitation is because of what I already commented in all the previous reviews. This contract allows only one buyback setup per user (address), this is why I think you are letting the user to change anything, because in other case, the user will have to use another account for doing another buyback with another setup. Using an ID, instead of the address, and allowing to create any number of immutable buy backs for a user, feels for natural in my opinion.

@samparsky  let me know if I'm mistaken on my previous assumptions

samparsky commented:
@anxolin Your comments are right, I am almost done with making the changes you suggested.

For number 3, yeah you should be able to get your money back after maybe a set time defined by the user or a defined time by the contract e.g. 3 months

Would that be okay?

samparsky commented:
@anxolin I am done making the changes regarding your comments

1. its not possible to break your promise anymore as all buybacks are immutable
2. The tip is enforced, its required before creating the buyback
3. A user is allowed to releaseFunds of a buyback if it expires. The minimum expiry time is one month. So a user can call `releaseBuybackFunds(buybackid)` after the buyback expires to have access to the funds
4. It's no longer possible to modify buybacks as you can create multiple buybacks.

Also the contract has been upgraded to 0.5.2 and the dx-contracts to the latest version. Looking forward to your comment.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+    // mapping of user address to ether deposit value
+    mapping(address => uint) public etherBalance;
+
+    // mapping of buyback id to auction index to ductchx index for the auction index participated in
```
The comment it's a bit confusing

samparsky replied:
Fixed

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+        address _userAddress = msg.sender;
+
+        require(hasEnoughDeposit(_auctionAmounts, _sellToken, _userAddress), "user does not have enough deposit to create buyback");
+        require(hasEnoughTippingDeposit(_userAddress, _tipAmount), "user does not have enough tip balance");
```
But, if the user adds for example:

```js
{
  ...
  _auctionIndexes: [35, 50],
  _auctionAmounts: [10, 10],
  _tipAmount: 5
}
```

- If the user deposits for tipping `5`, It will allow him to create the Buyback. 
- If a user executes the buyback for auction 35, it will get the tip
- If a user executes the buyback for auction 50, it will REVERT, cause there's not tip. This will break the promise and make the user to loose gas

samparsky replied:
Fixed to allowing the user to have only one index per buyback since multiple buybacks can be created now

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+     * @notice updateDutchExchange
+     */
+    function updateDutchExchange(DutchExchange _dx) external onlyOwner {
+        dx = _dx;
```
Security wise, this one is very sensitive. 

The owner could replace the DutchX with a contract that implements the deposit and the sell, but steals the tokens.

Do you see any solution to mitigate this?

samparsky replied:
The assumption would be that solidity should prevent updating it with anything that doesn't have the DuctchExchange abi

samparsky replied:
i have removed the function

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+            approveDutchX(sellToken, amount);
+            depositDutchx(sellToken, amount);
+
+            (dxIndex, newBuyerBalance) = dx.postSellOrder(sellToken, buyToken, auctionIndexes[i], amount);
```
When we execute a buy back, we sell for ALL the auctions in the buy back?

For example:
```js
{
  ...
  _auctionIndexes: [35, 50]
}
```

If the current auction is 35, it'll try to sell in number 35 (that it's all right), but then It'll try to sell in the 50, and it'll revert.

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+    }
+
+    /**
+     * @notice approve trading
```
Not sure what this comment means

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+        require(userBuyback.hasExecutedBuyback == false, "can only release funds of unexecuted buyback");
+        require(userBuyback.auctionIndexes.length > 0, "user does not exist");
+        
+        if(userBuyback.expires < block.timestamp) {
```
Any reason why here you didn't use the function `isTimePassed`?

anxolin commented on gnosis-buyback/contracts/Buyback.sol:
```diff
@@ -0,0 +1,539 @@
+     * @notice removeBuyBack
+     * Allows releasing expired buyback funds
+     */
+    function releaseBuyBackFund(uint _buybackId) public {
```
For me, what it make sense here is to release the auctions that has passed.

- You know the current auction for the token pair
- You can release the amounts for past auctions if no-one triggered the `postSellOrder`

samparsky replied:
The only issue here is when a user specifies 0 as the auctionIndex it becomes impossible to releaseFunds for that buyback

I added a condition to not allow 0 auctionIndex

samparsky commented:
@anxolin I have made changes to the effect of your comments and have responded to your review above.

Each buyback is a single buyback now which addresses a number of issues raised above i.e.
```
{
  ...
  _auctionIndex: 35
auctionAmount: 20
}
```

samparsky commented:
Hello @anxolin, I am hoping you can review the new changes asap

samparsky commented:
@anxolin Hello

samparsky commented:
@anxolin Welcome to the new month :)

samparsky commented:
@anxolin Hello

samparsky commented:
@anxolin Hello
